syntax = "proto3";

package posts;


service Posts {
    rpc CreatePost(Post) returns (PostResponse) {}

    rpc UpdatePost(UpdateRequest) returns (PostResponse) {}

    rpc DeletePost(GetDeleteRequest) returns (PostResponse) {}

    rpc GetPost(GetDeleteRequest) returns (GetPostResponse) {}

    rpc ListPosts(ListRequest) returns (PostList) {}
}

message Post {
    optional string id = 1;
    optional uint64 creator_id = 2;
    optional string text = 3;
    optional uint64 created_at = 4;
    optional uint64 edited_at = 5;
}

message PostList {
    repeated Post posts = 1;
}

message UpdateRequest {
    string post_id = 1;
    string text = 2;
    uint64 user_id = 3;
}

message GetDeleteRequest {
    string post_id = 1;
    uint64 user_id = 2;
}

message ListRequest {
    uint32 page = 1;
    uint64 user_id = 2;
}

message PostResponse {
    int32 http_status = 1;
    string message = 2;
    string post_id = 3;
}

message GetPostResponse {
    int32 http_status = 1;
    string message = 2;
    Post post = 3;
}
